src/ir/child-typer.h

```diff
@@ -0,0 +1,1074 @@
+      case GeSVecI64x2:
+      case EqVecF32x4:
+      case NeVecF32x4:
+      case LtVecF32x4:
+      case LeVecF32x4:
+      case GtVecF32x4:
+      case GeVecF32x4:
+      case EqVecF64x2:
+      case NeVecF64x2:
+      case LtVecF64x2:
+      case LeVecF64x2:
+      case GtVecF64x2:
+      case GeVecF64x2:
+      case AndVec128:
+      case OrVec128:
+      case XorVec128:
+      case AndNotVec128:
+      case AddVecI8x16:
+      case AddSatSVecI8x16:
+      case AddSatUVecI8x16:
+      case SubVecI8x16:
+      case SubSatSVecI8x16:
+      case SubSatUVecI8x16:
+      case MinSVecI8x16:
+      case MinUVecI8x16:
+      case MaxSVecI8x16:
+      case MaxUVecI8x16:
+      case AvgrUVecI8x16:
+      case Q15MulrSatSVecI16x8:
+      case ExtMulLowSVecI16x8:
+      case ExtMulHighSVecI16x8:
+      case ExtMulLowUVecI16x8:
+      case ExtMulHighUVecI16x8:
+      case AddVecI16x8:
+      case AddSatSVecI16x8:
+      case AddSatUVecI16x8:
+      case SubVecI16x8:
+      case SubSatSVecI16x8:
+      case SubSatUVecI16x8:
+      case MulVecI16x8:
+      case MinSVecI16x8:
+      case MinUVecI16x8:
+      case MaxSVecI16x8:
+      case MaxUVecI16x8:
+      case AvgrUVecI16x8:
+      case AddVecI32x4:
+      case SubVecI32x4:
+      case MulVecI32x4:
+      case MinSVecI32x4:
+      case MinUVecI32x4:
+      case MaxSVecI32x4:
+      case MaxUVecI32x4:
+      case DotSVecI16x8ToVecI32x4:
+      case ExtMulLowSVecI32x4:
+      case ExtMulHighSVecI32x4:
+      case ExtMulLowUVecI32x4:
+      case ExtMulHighUVecI32x4:
+      case AddVecI64x2:
+      case SubVecI64x2:
+      case MulVecI64x2:
+      case ExtMulLowSVecI64x2:
+      case ExtMulHighSVecI64x2:
+      case ExtMulLowUVecI64x2:
+      case ExtMulHighUVecI64x2:
+      case AddVecF32x4:
+      case SubVecF32x4:
+      case MulVecF32x4:
+      case DivVecF32x4:
+      case MinVecF32x4:
+      case MaxVecF32x4:
+      case PMinVecF32x4:
+      case PMaxVecF32x4:
+      case RelaxedMinVecF32x4:
+      case RelaxedMaxVecF32x4:
+      case AddVecF64x2:
+      case SubVecF64x2:
+      case MulVecF64x2:
+      case DivVecF64x2:
+      case MinVecF64x2:
+      case MaxVecF64x2:
+      case PMinVecF64x2:
+      case PMaxVecF64x2:
+      case RelaxedMinVecF64x2:
+      case RelaxedMaxVecF64x2:
+      case NarrowSVecI16x8ToVecI8x16:
+      case NarrowUVecI16x8ToVecI8x16:
+      case NarrowSVecI32x4ToVecI16x8:
+      case NarrowUVecI32x4ToVecI16x8:
+      case SwizzleVecI8x16:
+      case RelaxedSwizzleVecI8x16:
+      case RelaxedQ15MulrSVecI16x8:
+      case DotI8x16I7x16SToVecI16x8:
+        note(&curr->left, Type::v128);
+        note(&curr->right, Type::v128);
+        break;
+      case InvalidBinary:
+        WASM_UNREACHABLE("invalid binary op");
+    }
+  }
+
+  void visitSelect(Select* curr, std::optional<Type> type = std::nullopt) {
+    if (type) {
+      note(&curr->ifTrue, *type);
+      note(&curr->ifFalse, *type);
+    } else {
+      noteAny(&curr->ifTrue);
+      noteAny(&curr->ifFalse);
```

Comment:
This doesn't capture that they must be the same type. Is that not important when popping?

Reply:
They don't need to be the same type in general. In the validator, invalid cases like having children typed i32 and f32 should be caught because the result type would not be the LUB of the children types (unless it is `Type::none`, in which case it would fail another check that the result should be concrete).